### ISSUE TITLE
AS3 generating Math.POSITIVE_INFINITY, should be Number.POSITIVE_INFINITY

### ISSUE THREAD

**Issue by jcward** [HaxeFoundation/haxe#5882]
The AS3 generator is transpiling this Haxe:

```
    var p:Float = Math.POSITIVE_INFINITY;
    var n:Float = Math.NEGATIVE_INFINITY;
```

Into this AS3:

```
    var p : Number = Math["POSITIVE_INFINITY"];
    var n : Number = Math["NEGATIVE_INFINITY"];
```

However, the AS3 [Math](http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/Math.html) class doesn't have a POSITIVE_INFINITY or NEGATIVE_INFINITY property, so those Numbers become `NaN`.

Instead it should translate those properties to the [Number](http://help.adobe.com/en_US/FlashPlatform/reference/actionscript/3/Number.html) class:

```
    var p : Number = Number["POSITIVE_INFINITY"];
    var n : Number = Number["NEGATIVE_INFINITY"];
```

FYI, failed fix attempt at #5881 41b2e1e, later reverted by b8558a6

**Comment by jcward** [comment]
Also, why isn't this unit test catching it?  https://github.com/HaxeFoundation/haxe/blob/development/tests/unit/src/unitstd/Math.unit.hx#L219-L220

Is it being run and observed for AS3?

**Comment by Simn** [comment]
I'm not sure what you're actually reporting here. Can you post a minimal example of something that doesn't work as you expect?

**Comment by jcward** [comment]
Decorated via __main__.as or Boot.as. Closing.

**Comment by jcward** [comment]
Hi @ncannasse . I noticed you reopened this issue. Are you interested in the user experience around this issue? @Simn noted that the `__main__.as` class is intended as the app entry point. So I was able to find that `Boot.init()` sets up some `Array.prototype` stuff, and `__main__.as` sets up some monkey patches on the Math class.

It all works as expected, once I made sure these two AS3 language monkey patches were being called. As part of our build flow, I ended up post-processing the `__main__.as` class into a static helper, and calling it from the beginning of our app's Main entry point:

![selection_999 169](https://cloud.githubusercontent.com/assets/2192439/21199579/532336e8-c201-11e6-9231-23b4043daba5.png)

`Boot.init()` gets called statically, it looks like, from the code at the end of the file:

![image](https://cloud.githubusercontent.com/assets/2192439/21199671/b745af7a-c201-11e6-93dc-9300653fe18e.png)

Related issue: To get it to build, I also have to patch these two lines in `Boot.as`:

```
238: var arr : Array = this;
273: var str : String = this;
```

to:

```
238: var arr : Array = this as Array;
273: var str : String = this as String;
```

Otherwise the compiler (mxmlc) complains:

```
generated_as3/flash/Boot.as:238
Error: Implicit coercion of a value with static type Object to a possibly unrelated type Array.
				var arr : Array = this;
				                  ^

generated_as3/flash/Boot.as:273
Error: Implicit coercion of a value with static type Object to a possibly unrelated type String.
				var s : String = this;
				                 ^
```
